ISISScientificComputing/autoreduce
# End of run monitor restart

### Changes
* Add step into the end of run monitor windows service to allow for a restart of the service. 

This solution may have to be improved if we find that we are missing runs because of it. This could be a side effect of there being a file event (last_run.txt is updated) whilst the health check is taking place. If this is the case then we will have to have a separate Thread that monitors the service instead - this may infact be a separate windows service?
In order to do this the health check will have to be completely encapsulated from the end of run monitor and be purely external. If this can be incorporated into the implementation of #203 this would make the implementation of a separate windows service for monitoring much easier. 

### Test
* Point the end of run monitor to the data archive (overwrite `INST FOLDER` with  `INST_FOLDER = r'\\isis\inst$\NDX%s\Instrument'` in `monitors/settings.py`)
* Reinstall the end of run monitor - from command line (admin) type:
`python monitor_win_service.py install`
* In the Services tool for windows (type `Services` in start menu) locate the `Autoreduce Instrument Monitor` service.
* You will need to update the `Log On As` to an account with permission to view the archive
  * Right click the service and click properties
  * Navigate to the `Log On` tab
  * Check the `This account` section and put `clrc\<fed-id>` into the field and your password below
  * Click OK
* Right click the service again and click Start
* Once running go and check the `monitor.log` file located in `git-repo/logs/`
* You should see the standard start up text:
```
2018-10-18 09:57:26,992 Connecting to ActiveMQ...
2018-10-18 09:57:26,992 Connected to ActiveMQ
2018-10-18 09:57:27,005 Watching \\isis\inst$\NDXWISH\Instrument\logs
2018-10-18 09:57:27,019 Watching \\isis\inst$\NDXGEM\Instrument\logs
2018-10-18 09:57:27,026 Watching \\isis\inst$\NDXOSIRIS\Instrument\logs
2018-10-18 09:57:27,032 Watching \\isis\inst$\NDXPOLARIS\Instrument\logs
2018-10-18 09:57:27,039 Watching \\isis\inst$\NDXMUSR\Instrument\logs
2018-10-18 09:57:27,063 Watching \\isis\inst$\NDXPOLREF\Instrument\logs
```
* Wait around 15 seconds (current time out is set to 10 seconds but  interrupts from Wait for signal could cause slight delay)
* Refresh the log file (close and re-open)
* You should now see (obvious with the correct time):
```
2018-10-18 10:15:42,680 Performing Health check at 2018-10-18 10:15:42.681000
2018-10-18 10:15:42,680 No problems detected with service
```
* **Don't forget to the stop the service once you have finished testing else it will continue to poll the archive on your machine until it is told to stop!!!**
